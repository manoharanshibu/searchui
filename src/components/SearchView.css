.searchinput {
    color: #fff;
    background-color: #00008f;
    width: 150px;
    height: 40px;
    cursor: pointer;
    border: none;
    font-size: 14px;
    letter-spacing: 1px;
    line-height: 18px;
    font-family: "Source Sans Pro";
    font-weight: 600;
    text-align: center;
}

.container {
    width: 100%;
}

.search-container {
    margin-top: 30px;
    width: 100%;
    font-family: "Source Sans Pro";
    font-size: 14px;
    letter-spacing: 0;
    line-height: 17px;
    font-weight: 400;
}

.radio-container {
    display: inline-flex;
    margin: 30px 20px;
    width: 30%;
    font-family: "Source Sans Pro";
    font-size: 14px;
    letter-spacing: 0;
    line-height: 17px;
    font-family: "Source Sans Pro";
    font-weight: 400;
}

.searchicon {
    background-color: transparent;
    cursor: pointer;
    margin-right: 40px;
}

.searchicon:focus {
    outline: 0;
}