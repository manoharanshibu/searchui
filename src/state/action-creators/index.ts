import { ActionType } from "../action-types"
import { Action } from "../actions"
import { Dispatch } from 'redux';

export const fetchItems = () => {
    return async (dispatch: Dispatch<Action>) => {
        const columns = [
            { field: 'id', headerName: 'ID', width: 100 },
            { field: 'author', headerName: 'Author', width: 230 },
            { field: 'bookName', headerName: 'Book Name', width: 230 },
            { field: 'datePublished', headerName: 'Date Published', width: 180 },
            { field: 'bookDetails', headerName: 'Book Details', width: 300 },
        ];

        const rows = [
            { id: 1, bookName: 'Snow', author: 'Jon', datePublished: '01-01-2021', bookDetails: 'sample text' },
            { id: 2, bookName: 'Lannister', author: 'Cersei', datePublished: '01-02-2021', bookDetails: 'sample text' },
            { id: 3, bookName: 'Lannister', author: 'Jaime', datePublished: '05-01-2021', bookDetails: 'sample text' },
            { id: 4, bookName: 'Stark', author: 'Arya', datePublished: '01-01-2021', bookDetails: 'sample text' },
            { id: 5, bookName: 'Taralert(garyen', author: 'Daenerys', datePublished: '01-01-2021', bookDetails: 'sample text' },
            { id: 6, bookName: 'Melisandre', author: null, datePublished: '07-01-2020', bookDetails: 'sample text' },
            { id: 7, bookName: 'Clifford', author: 'Ferrara', datePublished: '01-01-2021', bookDetails: 'sample text' },
            { id: 8, bookName: 'Frances', author: 'Rossini', datePublished: '01-05-2020', bookDetails: 'sample text' },
            { id: 9, bookName: 'Roxie', author: 'Harvey', datePublished: '01-02-2021', bookDetails: 'sample text' },
        ];

        dispatch({
            type: ActionType.GET_ITEMS_SUCCESS,
            payload: {
                rows: rows,
                columns: columns
            }
        });
    }
}
