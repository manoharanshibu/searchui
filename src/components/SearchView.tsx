import * as React from "react";
import { useEffect, useState, useRef } from "react";
import TableComponent from './TableComponent';

import './SearchView';
import { SearchInput } from './SearchInput';

import { State } from '../state';
import { useDispatch, useSelector } from "react-redux";

import { fetchItems } from "../state/action-creators";

import Radio from '@mui/material/Radio';
import RadioGroup from '@mui/material/RadioGroup';
import FormControlLabel from '@mui/material/FormControlLabel';
import FormControl from '@mui/material/FormControl';

const SearchView = () => {
  const dispatch = useDispatch();
  const inputField = useRef();
  const { items, loadedItems } = useSelector((state: State) => state.itemsReducer);

  const [searchedData, setSearchedData] = useState<any>({ rows: [], columns: [] });
  const [selected, setSelected] = useState('author')

  useEffect(() => {
    dispatch(fetchItems());
  }, [dispatch]);

  useEffect(() => {
    setSearchedData(items)
  }, [loadedItems, items])

  const handleOnChange = (event: any) => {
    const _items = items.rows;
    if (event.target.value === '') {
      setSearchedData(items);
      return;
    }
    const data = _items.filter((item: any) => {
      if (selected === 'book') {
        return item && item.bookName ? item.bookName.toLowerCase().startsWith(event.target.value.toLowerCase()) : false;
      } else if (selected === 'date') {
        return item && item.datePublished ? new Date(item.datePublished).getTime() === new Date(event.target.value).getTime() : false
      } else {
        return item && item.author ? item.author.toLowerCase().startsWith(event.target.value.toLowerCase()) : false
      }
    })
    setSearchedData({
      columns: items.columns,
      rows: data
    })
  }


  return (
    <div className="container">
      <div className="search-container">
        <SearchInput
          searchedData={searchedData}
          inputField={inputField}
          handleOnChange={handleOnChange}
        />
      </div>
      <div className="radio-container">
        <FormControl component="fieldset">
          <RadioGroup
            aria-label="search"
            defaultValue="author"
            name="radio-buttons-group"
            onChange={(e: any) => setSelected(e.target.value)}
          >
            <FormControlLabel value="author" control={<Radio />} label="Search By Author" />
            <FormControlLabel value="book" control={<Radio />} label="Search By Book Name" />
            <FormControlLabel value="date" control={<Radio />} label="Search By Date" />
          </RadioGroup>
        </FormControl>
      </div>
      <TableComponent rows={searchedData.rows} columns={searchedData.columns} />
    </div>
  )
}

export default SearchView;
