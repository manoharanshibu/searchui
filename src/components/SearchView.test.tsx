import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import SearchView from './SearchView';

import { Provider } from 'react-redux'
import { store } from '../state/index';

describe(('SearchView Component'), () => {

    const initialState = {
        items: [],
        loadedItems: false
    }

    jest.mock('react-redux', () => ({
        useDispatch: () => { },
        useSelector: () => ({
            state: initialState,
        }),
    }));

    it('renders correctly', () => {
        const tree = renderer
            .create(<Provider store={store}><SearchView /></Provider>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
})
