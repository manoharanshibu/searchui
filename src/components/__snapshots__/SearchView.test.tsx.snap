// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchView Component renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="search-container"
  >
    <div
      className="search-input"
    >
      <div
        className="input-group"
      >
        <input
          aria-label="What are you searching for?"
          className="form-control border-right-0 border search-input-element"
          height="40px"
          onChange={[Function]}
          placeholder="What are you searching for?"
          type="search"
        />
        <span
          className="input-group-append searchicon"
        >
          <div
            className="input-group-text bg-transparent"
          >
            <i
              className="fa fa-search"
            />
          </div>
        </span>
      </div>
    </div>
  </div>
  <div
    className="radio-container"
  >
    <fieldset
      className="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
    >
      <div
        aria-label="search"
        className="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
        role="radiogroup"
      >
        <label
          className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            className="MuiRadio-root MuiRadio-colorPrimary MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root Mui-checked css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <input
              checked={true}
              className="PrivateSwitchBase-input css-1m9pwf3"
              disabled={false}
              name="radio-buttons-group"
              onChange={[Function]}
              type="radio"
              value="author"
            />
            <span
              className="css-hyxlzm"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                data-testid="RadioButtonUncheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-11zohuh-MuiSvgIcon-root"
                data-testid="RadioButtonCheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </span>
          </span>
          <span
            className="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            Search By Author
          </span>
        </label>
        <label
          className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            className="MuiRadio-root MuiRadio-colorPrimary MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <input
              checked={false}
              className="PrivateSwitchBase-input css-1m9pwf3"
              disabled={false}
              name="radio-buttons-group"
              onChange={[Function]}
              type="radio"
              value="book"
            />
            <span
              className="css-hyxlzm"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                data-testid="RadioButtonUncheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                data-testid="RadioButtonCheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </span>
          </span>
          <span
            className="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            Search By Book Name
          </span>
        </label>
        <label
          className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            className="MuiRadio-root MuiRadio-colorPrimary MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <input
              checked={false}
              className="PrivateSwitchBase-input css-1m9pwf3"
              disabled={false}
              name="radio-buttons-group"
              onChange={[Function]}
              type="radio"
              value="date"
            />
            <span
              className="css-hyxlzm"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                data-testid="RadioButtonUncheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                data-testid="RadioButtonCheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </span>
          </span>
          <span
            className="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            Search By Date
          </span>
        </label>
      </div>
    </fieldset>
  </div>
  <div
    className="tablecomponent"
  />
</div>
`;
